hudi-utilities/src/main/java/org/apache/hudi/utilities/streamer/StreamSync.java

```diff
@@ -574,12 +574,12 @@ private InputBatch fetchNextBatchFromSource(Option<String> resumeCheckpointStr,
       checkpointStr = dataAndCheckpoint.getCheckpointForNextBatch();
       if (this.userProvidedSchemaProvider != null && this.userProvidedSchemaProvider.getTargetSchema() != null
           && this.userProvidedSchemaProvider.getTargetSchema() != InputBatch.NULL_SCHEMA) {
+        // Let's deduce the schema provider for writer side first!
+        schemaProvider = getDeducedSchemaProvider(this.userProvidedSchemaProvider.getTargetSchema(), this.userProvidedSchemaProvider, metaClient);
```

Comment:
so, does this mean that all of row writer code path in StreamSync is now supported w/ our OOB schema evol ? 
Do we have tests for them ?

Reply:
Not necessarily. But adding this definitely won't make things worse. Previously, we were just using the schema provider without validating the schema at all, so even if the validation isn't strict enough, it is still better than no validation

Reply:
If you need more convincing, you can see at line 616 that we were already using deduced schema for row writer in that case. So we were already being inconsistent with when we use deduce. Additionally, in spark sql writer, all schemas go through deduce, even when bulk insert row writer is used there. So not promising that bulk insert row writer will have schema evolution working perfectly, but at least we will be consistent now. 

Deduce also fixes null schemas so they don't cause NPE. That's why my change at 610-613 is needed. The changes at 578-580 I will fight less hard for, but I think this will come back to haunt us later if we don't make that change now.

Reply:
gotcha. sg

Reply:
btw, do we have tests for flows that you fixed?